civiform/civiform
# Correct path for applicant_date_of_birth and fix existing data

### Description

This PR updates the "Well known path" for `applicant_date_of_birth` to include the nested `date` field.

This PR also adds a Durable Job that will update `applicant_date_of_birth` in all existing Applicant Data objects to include the nested `date` field. Although this Durable Job only really needs to run once, I utilized the existing infrastructure to set it up as recurring. Once this is merged, I'll plan to wait a month to allow plenty of time for the job to run and then I'll remove the job, the deprecated data path, and the conditional logic to handle both data paths.

### Checklist

#### General


### Instructions for manual testing

1. Check out the main branch, log in as a TI, and add a client
2. Check out this branch, log in as a TI, and add a client
3. Check that DOBs for both clients are showing up in the TI client list view
4. Log in as a program admin and create a program that includes the seeded "Applicant Date of Birth" question
5. (Optional) Apply as a TI on behalf of the first client you added. The answer to the "date of birth" question should be blank and attempting to edit it should result in an error.
6. Stop the server and edit the code so the new job will run on server start:
Add this to `RecurringJobExecutionTimeResolvers.java`
```
public static final class Immediately implements RecurringJobExecutionTimeResolver {
    @Override
    public Instant resolveExecutionTime(Clock clock) {
      return LocalDateTime.now(clock)
          .toInstant(ZoneId.systemDefault().getRules().getOffset(Instant.now()));
    }
  }
```
In `DurableJobModule.java`, edit the registration of the `FIX_APPLICANT_DOB_DATA_PATH` job to:
```
    durableJobRegistry.register(
        DurableJobName.FIX_APPLICANT_DOB_DATA_PATH,
        persistedDurableJob -> new FixApplicantDobDataPathJob(persistedDurableJob),
        new RecurringJobExecutionTimeResolvers.Immediately());
``` 
7. Restart the server and wait a minute for the job to run
8. Apply as a TI on behalf of the first client you added. The answer to the "date of birth" question should be filled in and you should be able to edit it.

### Issue(s) this completes

Fixes #5901
